src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs

```diff
@@ -804,8 +804,21 @@ protected bool TakeStartLine(SocketInput input)
                 QueryString = queryString;
                 HttpVersion = httpVersion;
 
-                bool caseMatches;
+                if (requestUrlPath.Length > 0 && requestUrlPath[0] != '/')
+                {
+                    Uri uri;
+                    if (Uri.TryCreate(requestUrlPath, UriKind.Absolute, out uri))
```

Comment:
System.Uri can't be used, it does lots of its own escaping and un-escaping.  I think your three interesting cases are:
- Starts with / if it's a path
- * for OPTIONS
- starts with http:// or https:// if its absolute, skip to the first / or ? after the scheme separator